sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/DoFnInvokers.java

```diff
@@ -0,0 +1,460 @@
+import net.bytebuddy.dynamic.scaffold.InstrumentedType;
+import net.bytebuddy.dynamic.scaffold.subclass.ConstructorStrategy;
+import net.bytebuddy.implementation.Implementation;
+import net.bytebuddy.implementation.MethodCall;
+import net.bytebuddy.implementation.bind.MethodDelegationBinder;
+import net.bytebuddy.implementation.bind.annotation.TargetMethodAnnotationDrivenBinder;
+import net.bytebuddy.implementation.bytecode.ByteCodeAppender;
+import net.bytebuddy.implementation.bytecode.StackManipulation;
+import net.bytebuddy.implementation.bytecode.Throw;
+import net.bytebuddy.implementation.bytecode.assign.Assigner;
+import net.bytebuddy.implementation.bytecode.member.FieldAccess;
+import net.bytebuddy.implementation.bytecode.member.MethodInvocation;
+import net.bytebuddy.implementation.bytecode.member.MethodReturn;
+import net.bytebuddy.implementation.bytecode.member.MethodVariableAccess;
+import net.bytebuddy.jar.asm.Label;
+import net.bytebuddy.jar.asm.MethodVisitor;
+import net.bytebuddy.jar.asm.Opcodes;
+import net.bytebuddy.matcher.ElementMatchers;
+
+import java.lang.reflect.Constructor;
+import java.lang.reflect.InvocationTargetException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.EnumMap;
+import java.util.LinkedHashMap;
+import java.util.Map;
+
+import javax.annotation.Nullable;
+
+/** Dynamically generates {@link DoFnInvoker} instances for invoking a {@link DoFn}. */
+public abstract class DoFnInvokers {
+  /**
+   * A cache of constructors of generated {@link DoFnInvoker} classes, keyed by {@link DoFn} class.
+   * Needed because generating an invoker class is expensive.
+   */
+  private static final Map<Class<?>, Constructor<?>> BYTE_BUDDY_INVOKER_CONSTRUCTOR_CACHE =
+      new LinkedHashMap<>();
+
+  private static final String FN_DELEGATE_FIELD_NAME = "delegate";
+
+  /** This is a factory class that should not be instantiated. */
+  private DoFnInvokers() {}
+
+  /**
+   * If this is an {@link OldDoFn} produced via {@link #toOldDoFn}, returns the class of the
+   * original {@link DoFn}, otherwise returns {@code fn.getClass()}.
+   */
+  public static Class<?> getDoFnClass(OldDoFn<?, ?> fn) {
+    if (fn instanceof DoFnAdapters.SimpleDoFnAdapter) {
+      return ((DoFnAdapters.SimpleDoFnAdapter<?, ?>) fn).fn.getClass();
+    } else {
+      return fn.getClass();
+    }
+  }
+
+  /** Create a {@link OldDoFn} that the {@link DoFn}. */
+  public static <InputT, OutputT> OldDoFn<InputT, OutputT> toOldDoFn(DoFn<InputT, OutputT> fn) {
+    DoFnSignature signature = DoFnReflector.getSignature(fn.getClass());
+    if (signature.getProcessElement().usesSingleWindow()) {
+      return new DoFnAdapters.WindowDoFnAdapter<>(fn);
+    } else {
+      return new DoFnAdapters.SimpleDoFnAdapter<>(fn);
+    }
+  }
+
+  /** @return the {@link DoFnInvoker} for the given {@link DoFn}. */
+  public static <InputT, OutputT> DoFnInvoker<InputT, OutputT> newByteBuddyInvoker(
+      DoFn<InputT, OutputT> fn) {
```

Comment:
I think this could take the signature, right? The signature should know the function it is a signature for. That would also mean that this class doesn't need to get the signature from a function.